[jit] bind autograd.grad function into TorchScript

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23913 [jit] bind autograd.backward and tensor.backward in TorchScript
* **#24871 [jit] bind autograd.grad function into TorchScript**

Summary:
Bind the torch.autograd.grad function into TorchScript so that well
formed inputs can directly call this from a TorchScript function.

This also change the serliazation a bit, it fixes a small bug where node
output type can never be tensor type in prim::ListConstruct(only its elementype can be), and add the case where we need to annotate the ListType if the element type is optional type to preserve type information when reimport

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24871

Differential Revision: [D16923273](https://our.internmc.facebook.com/intern/diff/D16923273)

wanchaol committed:
[jit] bind autograd.grad function into TorchScript

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] bind autograd.grad function into TorchScript"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] bind autograd.grad function into TorchScript"


Summary:
Bind the torch.autograd.grad function into TorchScript so that well
formed inputs can directly call this from a TorchScript function.

This also change the serliazation a bit, it fixes a small bug where node
output type can never be tensor type in prim::ListConstruct(only its elementype can be), and add the case where we need to annotate the ListType if the element type is optional type to preserve type information when reimport

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24871

wanchaol committed:
Update on "[jit] bind autograd.grad function into TorchScript"


Summary:
Bind the torch.autograd.grad function into TorchScript so that well
formed inputs can directly call this from a TorchScript function.

This also change the serliazation a bit, it fixes a small bug where node
output type can never be tensor type in prim::ListConstruct(only its elementype can be), and add the case where we need to annotate the ListType if the element type is optional type to preserve type information when reimport

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24871

wanchaol committed:
Update on "[jit] bind autograd.grad function into TorchScript"


Summary:
Bind the torch.autograd.grad function into TorchScript so that well
formed inputs can directly call this from a TorchScript function.

This also change the serliazation a bit, it fixes a small bug where node
output type can never be tensor type in prim::ListConstruct(only its elementype can be), and add the case where we need to annotate the ListType if the element type is optional type to preserve type information when reimport

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24871

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -1010,15 +1010,30 @@ struct PythonPrintPass {
+            if (node->inputs().size() == 0) {
+              stmt << "annotate(" << node->output()->type()->python_str()
+                   << ", [])";
+              break;
```
These non-local breaks are hard to follow. Can you avoid them? It is easy to miss the fact that control ends here.

zdevito reviewed (approved):
Looks good. I have a few minor issues that should be addressed.

wanchaol committed:
address comments on "[jit] bind autograd.grad function into TorchScript"


Summary:
Bind the torch.autograd.grad function into TorchScript so that well
formed inputs can directly call this from a TorchScript function.

This also change the serliazation a bit, it fixes a small bug where node
output type can never be tensor type in prim::ListConstruct(only its elementype can be), and add the case where we need to annotate the ListType if the element type is optional type to preserve type information when reimport

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24871

Differential Revision: [D16923273](https://our.internmc.facebook.com/intern/diff/D16923273)
